### PR TITLE
[FLINK-947] Add a declarative expression API

### PR DESCRIPTION
**Pull request by aljoscha**
This one is quite big. So you should check out the documentation, skaldic, examples and test cases to see how the API works.

### CODE REVIEW

File: flink-staging/flink-linq/pom.xml [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+		<dependency>
+			<groupId>org.scala-lang</groupId>
+			<artifactId>scala-library</artifactId>
+		</dependency>
+
+		<dependency>
+			<groupId>org.scala-lang</groupId>
+			<artifactId>scala-compiler</artifactId>
+		</dependency>
+
```
Comment (mbalassi) [inline flink-staging/flink-linq/pom.xml]: You transitively depend on the scala stuff through flink-scala, so you could omit these.

**Comment by mbalassi** [comment]
Great additions! Really looking to trying it out. :)

File: flink-staging/flink-linq/pom.xml (hunk 1 above)
Comment (rmetzger) [inline flink-staging/flink-linq/pom.xml]: I think its really not an issue to directly add your dependencies to the pom.
Imagine we change something in the `flink-scala` module.
I was actually thinking about adding a check to maven that every dependency has to be added directly. I'm pretty sure there are cases in the project where we use stuff Apache Commons libraries which come from external dependencies.

File: docs/linq.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+KIND, either express or implied.  See the License for the
+specific language governing permissions and limitations
+under the License.
+-->
+
+* This will be replaced by the TOC
+{:toc}
+
+**Language-Integrated Queries are an experimental feature and can currently only be used with
```
Comment (rmetzger) [inline docs/linq.md]: Good to see some documentation as well!

Which types are supported by the expression API? Only scala case-classes? POJOs ? Even more? 
Would be could if you could add that to the documentation.

File: flink-staging/flink-linq/src/main/scala/org/apache/flink/api/scala/expressions/package.scala [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+ *   case class Word(word: String, count: Int)
+ *
+ *   val result = in.select(...).as('word, 'count)
+ *   val set = result.as[Word]
+ * }}}
+ */
+package object expressions extends ImplicitExpressionConversions {
+
+  implicit def dataSet2DataSetConversions[T](set: DataSet[T]): DataSetConversions[T] = {
+    Preconditions.checkArgument(set.getType.isInstanceOf[CompositeType[T]])
```
Comment (rmetzger) [inline flink-staging/flink-linq/src/main/scala/org/apache/flink/api/scala/expressions/package.scala]: I think we need a good error message here, telling the user why the type cannot be used with the expr lang.

**Comment by rmetzger** [comment]
Very very nice work! I've played around a bit with it and the first impression is very good.

+1 to merge the pull request. The change is very big, but stable enough to be merged to master. Not merging it soon would probably cause a lot of work on @aljoscha side.

**Comment by rmetzger** [comment]
One more thing, the maven module is called "flink-linq". Are we certain that we can use the name LINQ without problems here?

**Comment by aljoscha** [comment]
Yeah, I'm not sure about linq as well. I like the name but realise that it might be problematic. What do the others think. I could call it "flink-expressions".

I will add documentation about which types are supported and a good error message for unsupported types as @rmetzger mentioned.

**Comment by mxm** [comment]
To me, `flink-expressions` sounds much better than `linq` and it mitigates the risk of law suites :)

**Pull request merged**